File: MultiLanguageSandbox/src/test/deps/java/4.java
import java.util.ArrayList;
import java.util.List;
import java.lang.*;
import java.util.*;

class Solution {

  /**
   * Given an array of distinct integers, return all possible permutations of the array.
   * >>> permute(new int[]{1,2,3})
   * [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
   * >>> permute(new int[]{0,1})
   * [[0,1],[1,0]]
   * >>> permute(new int[]{1})
   * [[1]]
   */
  public List<List<Integer>> permute(int[] nums) 
{
        List<List<Integer>> result = new ArrayList<>();
        backtrack(nums, 0, result);
        return result;
    }
    
    private void backtrack(int[] nums, int start, List<List<Integer>> result) {
        if (start == nums.length - 1) {
            // Convert the current permutation to a list and add to result
            List<Integer> permutation = new ArrayList<>();
            for (int num : nums) {
                permutation.add(num);
            }
            result.add(permutation);
            return;
        }
        
        for (int i = start; i < nums.length; i++) {
            // Swap the current element with the start element
            swap(nums, start, i);
            // Recursively generate permutations for the remaining elements
            backtrack(nums, start + 1, result);
            // Backtrack (swap back to original position)
            swap(nums, start, i);
        }
    }
    
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
public static void main(String[] args) {
    Solution solution = new Solution();

    // Test the function with different test cases
    assert solution
      .permute(new int[] { 4, 5, 6 })
      .equals(
        Arrays.asList(
          Arrays.asList(4, 5, 6),
          Arrays.asList(4, 6, 5),
          Arrays.asList(5, 4, 6),
          Arrays.asList(5, 6, 4),
          Arrays.asList(6, 4, 5),
          Arrays.asList(6, 5, 4)
        )
      );
    assert solution
      .permute(new int[] { 7, 8 })
      .equals(Arrays.asList(Arrays.asList(7, 8), Arrays.asList(8, 7)));
    assert solution
      .permute(new int[] { 9 })
      .equals(Collections.singletonList(Collections.singletonList(9)));
    System.out.println("All tests passed");
  }
}

File: MultiLanguageSandbox/src/test/deps/dart/4.dart

List<List<int>> generatePermutations(List<int> list)
/*Generates all possible permutations of a given list of integers.

This recursive function takes a list of integers as input and returns
a list of lists, where each inner list represents a unique permutation
of the input list. The function works by iterating over each element in
the list, removing it, and generating all permutations of the remaining
elements. Each of these permutations has the removed element prepended
to it, thereby creating a new permutation that includes the removed element.
The base case of the recursion is when the list contains only one element,
in which case, the list itself is the only permutation.

Args:
  list (List<int>): The list of integers to permute.

Returns:
  List<List<int>>: A list containing all permutations of the input list.

Examples:
  generatePermutations([1]) returns [[1]]
  generatePermutations([1, 2]) returns [[1, 2], [2, 1]]
*/
{
  // Base case: if the list has only one element, return it as the only permutation
  if (list.length == 1) {
    return [List.from(list)];
  }

  List<List<int>> permutations = [];

  for (int i = 0; i < list.length; i++) {
    // Remove the current element
    int currentElement = list[i];
    List<int> remainingElements = List.from(list)..removeAt(i);

    // Recursively generate permutations of the remaining elements
    List<List<int>> subPermutations = generatePermutations(remainingElements);

    // Prepend the current element to each sub-permutation
    for (List<int> subPermutation in subPermutations) {
      subPermutation.insert(0, currentElement);
      permutations.add(subPermutation);
    }
  }

  return permutations;
}
void testGeneratePermutations() {
  // Test case 1
  List<int> list1 = [1];
  List<List<int>> expectedPerms1 = [
    [1]
  ];
  assert(ListEquality().equals(generatePermutations(list1), expectedPerms1));

  // Test case 2
  List<int> list2 = [1, 2];
  List<List<int>> expectedPerms2 = [
    [1, 2],
    [2, 1]
  ];
  assert(ListEquality().equals(generatePermutations(list2), expectedPerms2));

  // Test case 3
  List<int> list3 = [1, 2, 3];
  List<List<int>> expectedPerms3 = [
    [1, 2, 3],
    [1, 3, 2],
    [2, 1, 3],
    [2, 3, 1],
    [3, 1, 2],
    [3, 2, 1]
  ];
  assert(ListEquality().equals(generatePermutations(list3), expectedPerms3));

  print("All test cases passed!");
}

// Equality for nested lists
class ListEquality {
  bool equals(List<List<int>> list1, List<List<int>> list2) {
    if (list1.length != list2.length) return false;
    for (int i = 0; i < list1.length; i++) {
      if (!_listEquals(list1[i], list2[i])) {
        return false;
      }
    }
    return true;
  }

  bool _listEquals(List<int> list1, List<int> list2) {
    if (list1.length != list2.length) return false;
    for (int i = 0; i < list1.length; i++) {
      if (list1[i] != list2[i]) {
        return false;
      }
    }
    return true;
  }
}

void main() {
  testGeneratePermutations();
}